Add hoomd.version.floating_point_precision and summarize features

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should based on *maint*. -->
<!-- New features should based on *master*. -->

## Description

<!-- Describe your changes in detail. -->
Add hoomd.version.floating_point_precision.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->
Allow user scripts and plugins to determine what floating point precision used in the build.

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
Tested locally:
```bash
$ python3 -c "import hoomd; print(hoomd.version.floating_point_precision)"
(64, 32)
```
<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Added:

* ``hoomd.version.floating_point_precision`` - Floating point width in bits for the particle properties and local calculations.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

joaander commented:
@RainierBarrett or @whitead, please review and comment on whether this API fits your needs.

b-butler reviewed (changes requested):
Looks good to me. I would document that the second value only currently effects some values in HPMC and the first is for MD and particular data in HPMC.

cbkerr commented on hoomd/HOOMDVersion.cc:
```diff
@@ -173,4 +173,15 @@ std::string BuildInfo::getInstallDir()
+    {
+#if defined(SINGLE_PRECISION)
+    return std::make_pair(32, 32);
+#elif !defined(SINGLE_PRECISION) && defined(ENABLE_HPMC_MIXED_PRECISION)
+    return std::make_pair(64, 32);
+#elif !defined(SINGLE_PRECISION) && !defined(ENABLE_HPMC_MIXED_PRECISION)
+    return std::make_pair(64, 64);
```
Some of the conditions look redundant. Is that for being explicit?

```suggestion
#elif defined(ENABLE_HPMC_MIXED_PRECISION)
    return std::make_pair(64, 32);
#elif
    return std::make_pair(64, 64);
```

joaander replied:
I'm planning ahead. Per request by users in #355, all 4 options will possible when we enable MD mixed precision.

cbkerr reviewed (commented):
I can't comment on the use cases for this but see inline comment.

joaander commented:
> Looks good to me. I would document that the second value only currently effects some values in HPMC and the first is for MD and particular data in HPMC.

I'd prefer to keep the API docs for `floating_point_precision` clean and simple - else we will forget to update them in the future. I don't see an obvious place in the documentation to provide an overview of the mixed precision capabilities offered by the code. Do you think we should add a "Features" section to discuss this?

whitead commented:
Exactly what we need. Thanks!

b-butler commented:
@joaander  is the documentation saying that the first number represents storage and the second local calculations? If so, I didn't find it clean what each individual value in `floating_point_precision` represented, and being more explicit on what each index represents would be more helpful. We don't need to add HPMC specific documentation if we are trying to keep this the same when mixed precision MD is added.

b-butler commented:
Also, having a features section would be good that outlines the default behavior and available options.

joaander commented:
> Also, having a features section would be good that outlines the default behavior and available options.

In a future PR we could add this.

joaander commented:
> @joaander is the documentation saying that the first number represents storage and the second local calculations? If so, I didn't find it clean what each individual value in `floating_point_precision` represented, and being more explicit on what each index represents would be more helpful. We don't need to add HPMC specific documentation if we are trying to keep this the same when mixed precision MD is added.

I expanded the documentation. At this level, the docs can't possibly list every location where the two precisions are used. Perhaps in the mixed precision PR I could update the API docs for every operation and document when they implement mixed precision.

joaander commented:
I didn't like the long description in `version.py` so I added `features.rst` which summarizes all HOOMD-blue features including those affected by compile time options. I also updated the docs of those operations for selectively implemented features.

b-butler reviewed (approved):
Thanks, sorry for the extra work, but I think this will help user much better understand the values found in `hoomd.version.floating_point_precision`.

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+-------------------------
+
+HOOMD-blue can perform simulations of hard particles using the Monte Carlo method (`hpmc`). Hard
+particles are defined by their shape and the `HPMC integrator <hpmc.integrate>` supports
```
```suggestion
particles are defined by their shape, and the `HPMC integrator <hpmc.integrate>` supports
```

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+`md.update.ActiveRotationalDiffusion`. At runtime, `hoomd.version.md_built` indicates whether the
+build supports MD simulations.
+
+md.update
```
typo?

joaander replied:
Yes.

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+compile time with the ``ENABLE_GPU`` CMake option (see :doc:`building`). Select the device to use at
+run time with the `device <hoomd.device>` module. Unless otherwise stated in the documentation,
+**all** operations and methods support GPU execution. At runtime, `hoomd.version.gpu_enabled` indicates
+whether the build supports run time compilation.
```
Does run time compilation mean JIT code?

Or does it mean "able to run on GPU"?

```suggestion
whether the build supports running on the GPU.
```

joaander replied:
Fixed.

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+---------
+
+Some operations in HOOMD-blue can use multiple CPU threads in a single process. Control this with
+the `device.Device.num_cpu_threads` property. In this release, threading support in HOOMD-blue is
```
"in this release"

Is there a way to set a reminder to update this when that changes? Maybe just a code comment

joaander replied:
We just need to be diligent. There is no direct connection between what is written in the code and the docs. This may never change. At best I would hope that we remember to update the local docs for each operation when we change them. But the overview docs are the most likely to get out of date.

cbkerr replied:
maybe just tag with like "as of 3.0" so readers know when it was update?
```suggestion
the `device.Device.num_cpu_threads` property. As of version 3.0, threading support in HOOMD-blue is
```

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+the `device.Device.num_cpu_threads` property. In this release, threading support in HOOMD-blue is
+very limited and only applies to implicit depletants in `hpmc.integrate.HPMCIntegrator`, and
+`hpmc.pair.user.CPPPotentialUnion`. Threading must must be enabled at compile time with the
+``ENABLE_TBB`` CMake option (see :doc:`building`). At runtime, `hoomd.version.tbb_enabled` indicates
+whether the build supports MPI.
```
Reading this I'm confused about the discrepancy between the flag saying TBB and the description saying MPI again.

joaander replied:
Right, copy and paste bug. Fixed.

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+
+Some operations allow the user to provide arbitrary C++ code which HOOMD-blue compiles at run time
+and executes during the simulation. `hpmc.pair.user` and `hpmc.external.user` enable users to apply
+arbiratray pair and external potentials to particles in HPMC simulations. `hpmc.pair.user` is
```
```suggestion
arbitrary pair and external potentials to particles in HPMC simulations. `hpmc.pair.user`
```

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+Run time compilation
+--------------------
+
+Some operations allow the user to provide arbitrary C++ code which HOOMD-blue compiles at run time
```
```suggestion
Some operations allow the user to provide arbitrary C++ code that HOOMD-blue compiles at run time
```

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+Some operations allow the user to provide arbitrary C++ code which HOOMD-blue compiles at run time
+and executes during the simulation. `hpmc.pair.user` and `hpmc.external.user` enable users to apply
+arbiratray pair and external potentials to particles in HPMC simulations. `hpmc.pair.user` is
+supports both CPUs and NVIDIA GPUs while `hpmc.external.user` is only supports CPUs. Run time
```
```suggestion
supports both CPUs and NVIDIA GPUs while `hpmc.external.user` only supports CPUs. Run time
```

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+
+HOOMD-blue performs computations with mixed floating point precision. There is a **high precision**
+type and a **reduced precision** type. All particle properties are stored in the **high precision**
+type and most operations also perform all computations with **high precision**. Operations that to
```
```suggestion
type, and most operations also perform all computations with **high precision**. Operations that do
```

cbkerr commented on sphinx-doc/features.rst:
```diff
@@ -0,0 +1,139 @@
+---------------
+
+HOOMD-blue performs computations with mixed floating point precision. There is a **high precision**
+type and a **reduced precision** type. All particle properties are stored in the **high precision**
```
I would not bold the high/reduced precision terms after they are first mentioned. It provides too much emphasis. Perhaps use a code syntax style?

joaander replied:
I just removed the bold.

cbkerr reviewed (commented):
Small comments on the documentation. I will commit a batch of quick grammar fixes but some others remain.

cbkerr reviewed (approved):
Thanks for those tweaks, looks good now!
